// @redwoodjs/cli/dist/commands/dbCommands/up.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handler = exports.builder = exports.desc = exports.command = void 0;

var _lib = require("../../lib");

var _generate = require("./generate");

const command = 'up';
exports.command = command;
const desc = 'Generate the Prisma client and apply migrations.';
exports.desc = desc;
const builder = {
  verbose: {
    type: 'boolean',
    default: true,
    alias: ['v']
  },
  dbClient: {
    type: 'boolean',
    default: true
  }
};
exports.builder = builder;

const handler = async ({
  verbose = true,
  dbClient = true
}) => {
  const success = await (0, _lib.runCommandTask)([{
    title: 'Migrate database up...',
    cmd: 'yarn prisma',
    args: ['migrate up', '--experimental', '--create-db']
  }], {
    verbose
  });

  if (success && dbClient) {
    await (0, _generate.handler)({
      force: true,
      verbose
    });
  }
};

exports.handler = handler;
