// @redwoodjs/cli/dist/lib/test.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.loadFixture = exports.loadGeneratorFixture = exports.generatorsRootPath = void 0;

var _fs = _interopRequireDefault(require("fs"));

var _path = _interopRequireDefault(require("path"));

// Include at the top of your tests. Automatically mocks out the file system
//
// import { loadComponentFixture } from 'src/lib/test'
//
// test('true is true', () => {
//   expect('some output').toEqual(loadComponentFixture('component', 'filename.js'))
// })
jest.mock('@redwoodjs/internal', () => {
  const path = require('path');

  return { ...require.requireActual('@redwoodjs/internal'),
    getPaths: () => {
      const BASE_PATH = '/path/to/project';
      return {
        base: BASE_PATH,
        api: {
          db: path.join(global.__dirname, 'fixtures'),
          // this folder
          src: path.join(BASE_PATH, './api/src'),
          services: path.join(BASE_PATH, './api/src/services'),
          graphql: path.join(BASE_PATH, './api/src/graphql'),
          functions: path.join(BASE_PATH, './api/src/functions')
        },
        web: {
          src: path.join(BASE_PATH, './web/src'),
          routes: path.join(BASE_PATH, 'web/src/Routes.js'),
          components: path.join(BASE_PATH, '/web/src/components'),
          layouts: path.join(BASE_PATH, '/web/src/layouts'),
          pages: path.join(BASE_PATH, '/web/src/pages')
        }
      };
    }
  };
});

const generatorsRootPath = _path.default.join(__dirname, '..', 'commands', 'generate'); // Loads the fixture for a generator by assuming a lot of the path structure automatically:
//
//   loadGeneratorFixture('scaffold', 'NamePage.js')
//
// will return the contents of:
//
//   cli/src/commands/generate/scaffold/test/fixtures/NamePage.js.fixture


exports.generatorsRootPath = generatorsRootPath;

const loadGeneratorFixture = (generator, name) => {
  return loadFixture(_path.default.join(__dirname, '..', 'commands', 'generate', generator, '__tests__', 'fixtures', name));
}; // Returns the contents of a text file suffixed with ".fixture"


exports.loadGeneratorFixture = loadGeneratorFixture;

const loadFixture = filepath => {
  return _fs.default.readFileSync(filepath).toString();
};

exports.loadFixture = loadFixture;
