// @redwoodjs/cli/dist/commands/generate/function/function.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handler = exports.command = exports.builder = exports.desc = exports.files = void 0;

var _path = _interopRequireDefault(require("path"));

var _camelcase = _interopRequireDefault(require("camelcase"));

var _lib = require("../../../lib");

var _helpers = require("../helpers");

const files = async ({
  name,
  ...rest
}) => {
  const functionName = (0, _camelcase.default)(name);
  const file = (0, _helpers.templateForComponentFile)({
    name: functionName,
    componentName: functionName,
    apiPathSection: 'functions',
    generator: 'function',
    templatePath: 'function.js.template',
    templateVars: { ...rest
    },
    outputPath: _path.default.join((0, _lib.getPaths)().api.functions, `${functionName}.js`)
  });
  return {
    [file[0]]: file[1]
  };
};

exports.files = files;
const desc = 'Generate a function';
exports.desc = desc;
const builder = {
  force: {
    type: 'boolean',
    default: false
  }
};
exports.builder = builder;
const {
  command,
  handler
} = (0, _helpers.createYargsForComponentGeneration)({
  componentName: 'function',
  filesFn: files
});
exports.handler = handler;
exports.command = command;
