// @redwoodjs/cli/dist/commands/destroy/helpers.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.createYargsForComponentDestroy = void 0;

var _listr = _interopRequireDefault(require("listr"));

var _lib = require("../../lib");

var _colors = _interopRequireDefault(require("../../lib/colors"));

const tasks = ({
  componentName,
  filesFn,
  name
}) => new _listr.default([{
  title: `Destroying ${componentName} files...`,
  task: async () => {
    const f = await filesFn({
      name
    });
    return (0, _lib.deleteFilesTask)(f);
  }
}], {
  collapse: false,
  exitOnError: true
});

const createYargsForComponentDestroy = ({
  componentName,
  filesFn
}) => {
  return {
    command: `${componentName} <name>`,
    desc: `Destroy a ${componentName} component.`,
    handler: async ({
      name
    }) => {
      const t = tasks({
        componentName,
        filesFn,
        name
      });

      try {
        await t.run();
      } catch (e) {
        console.log(_colors.default.error(e.message));
      }
    },
    tasks
  };
};

exports.createYargsForComponentDestroy = createYargsForComponentDestroy;
