// @redwoodjs/api/dist/functions/graphQLServer.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.createGraphQLHandler = exports.createContextHandler = void 0;

var _apolloServerLambda = require("apollo-server-lambda");

var _authHeaders = require("../auth/authHeaders");

var _globalContext = require("../globalContext");

//

/**
 * We use Apollo Server's `context` option as an entry point for constructing our own
 * global context object.
 *
 * Context explained Apollo's Docs:
 * Context is an object shared by all resolvers in a particular query,
 * and is used to contain per-request state, including authentication information,
 * dataloader instances, and anything else that should be taken into account when
 * resolving the query.
 */
const createContextHandler = (userContext, getCurrentUser) => {
  return async ({
    event,
    context
  }) => {
    // Prevent the Lambda function from waiting for all resources,
    // such as database connections, to be released before returning a reponse.
    context.callbackWaitsForEmptyEventLoop = false; // Get the authorization information from the request headers and request context.

    const type = (0, _authHeaders.getAuthProviderType)(event);

    if (typeof type !== 'undefined') {
      const authToken = await (0, _authHeaders.decodeAuthToken)({
        type,
        event,
        context
      });
      context.currentUser = typeof getCurrentUser == 'function' ? await getCurrentUser(authToken) : authToken;
    }

    if (typeof userContext === 'function') {
      userContext = await userContext({
        event,
        context
      });
    } // Sets the **global** context object, which can be imported with:
    // import { context } from '@redwoodjs/api'


    return (0, _globalContext.setContext)({ ...context,
      ...userContext
    });
  };
};

exports.createContextHandler = createContextHandler;

/**
 * Creates an Apollo GraphQL Server.
 *
 * ```js
 * export const handler = createGraphQLHandler({ schema, context, getCurrentUser })
 * ```
 */
const createGraphQLHandler = ({
  context,
  getCurrentUser,
  onException,
  ...options
} = {}, db) => {
  const handler = new _apolloServerLambda.ApolloServer({
    playground: process.env.NODE_ENV !== 'production',
    ...options,
    context: createContextHandler(context, getCurrentUser)
  }).createHandler();
  return (event, context, callback) => {
    try {
      handler(event, context, callback);
    } catch (e) {
      onException && onException(); // Disconnect from the database (recommended by Prisma), this step will be
      // removed in future releases.

      db && db.disconnect();
      throw e;
    }
  };
};

exports.createGraphQLHandler = createGraphQLHandler;
