// @redwoodjs/cli/dist/commands/destroy/page/page.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handler = exports.tasks = exports.desc = exports.command = void 0;

var _camelcase = _interopRequireDefault(require("camelcase"));

var _listr = _interopRequireDefault(require("listr"));

var _lib = require("../../../lib");

var _colors = _interopRequireDefault(require("../../../lib/colors"));

var _helpers = require("../../generate/helpers");

var _page = require("../../generate/page/page");

const command = 'page <name> [path]';
exports.command = command;
const desc = 'Destroy a page component.';
exports.desc = desc;

const tasks = ({
  name,
  path
}) => new _listr.default([{
  title: 'Destroying page files...',
  task: async () => {
    const f = (0, _page.files)({
      name,
      path: (0, _helpers.pathName)(path, name)
    });
    return (0, _lib.deleteFilesTask)(f);
  }
}, {
  title: 'Cleaning up routes file...',
  task: async () => (0, _lib.removeRoutesFromRouterTask)([(0, _camelcase.default)(name)])
}], {
  collapse: false,
  exitOnError: true
});

exports.tasks = tasks;

const handler = async ({
  name,
  path
}) => {
  const t = tasks({
    name,
    path
  });

  try {
    await t.run();
  } catch (e) {
    console.log(_colors.default.error(e.message));
  }
};

exports.handler = handler;
