// @redwoodjs/cli/dist/commands/destroy/sdl/sdl.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handler = exports.tasks = exports.desc = exports.command = void 0;

var _listr = _interopRequireDefault(require("listr"));

var _lib = require("../../../lib");

var _colors = _interopRequireDefault(require("../../../lib/colors"));

var _sdl = require("../../generate/sdl/sdl");

const command = 'sdl <model>';
exports.command = command;
const desc = 'Destroy a GraphQL schema and service object.';
exports.desc = desc;

const tasks = ({
  model
}) => new _listr.default([{
  title: 'Destroying GraphQL schema and service object files...',
  task: async () => {
    const f = await (0, _sdl.files)({
      name: model,
      crud: false
    });
    return (0, _lib.deleteFilesTask)(f);
  }
}], {
  collapse: false,
  exitOnError: true
});

exports.tasks = tasks;

const handler = async ({
  model
}) => {
  const t = tasks({
    model
  });

  try {
    await t.run();
  } catch (e) {
    console.log(_colors.default.error(e.message));
  }
};

exports.handler = handler;
