// @redwoodjs/cli/dist/commands/generate/sdl/sdl.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handler = exports.builder = exports.desc = exports.command = exports.files = void 0;

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/concat"));

var _promise = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/promise"));

var _reduce = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/reduce"));

var _find = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/find"));

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/index-of"));

var _filter = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/filter"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/map"));

var _entries = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/object/entries"));

var _path = _interopRequireDefault(require("path"));

var _listr = _interopRequireDefault(require("listr"));

var _camelcase = _interopRequireDefault(require("camelcase"));

var _pascalcase = _interopRequireDefault(require("pascalcase"));

var _pluralize = _interopRequireDefault(require("pluralize"));

var _lib = require("../../../lib");

var _colors = _interopRequireDefault(require("../../../lib/colors"));

var _service = require("../service/service");

var _helpers = require("../helpers");

const IGNORE_FIELDS_FOR_INPUT = ['id', 'createdAt'];

const modelFieldToSDL = (field, required = true, types = {}) => {
  if ((0, _entries.default)(types).length) {
    field.type = field.kind === 'object' ? idType(types[field.type]) : field.type;
  }

  return `${field.name}: ${field.type}${field.isRequired && required ? '!' : ''}`;
};

const querySDL = model => {
  var _context;

  return (0, _map.default)(_context = model.fields).call(_context, field => modelFieldToSDL(field));
};

const inputSDL = (model, required, types = {}) => {
  var _context2, _context3;

  return (0, _map.default)(_context2 = (0, _filter.default)(_context3 = model.fields).call(_context3, field => {
    return (0, _indexOf.default)(IGNORE_FIELDS_FOR_INPUT).call(IGNORE_FIELDS_FOR_INPUT, field.name) === -1 && field.kind !== 'object';
  })).call(_context2, field => modelFieldToSDL(field, required, types));
}; // creates the CreateInput type (all fields are required)


const createInputSDL = (model, types = {}) => {
  return inputSDL(model, true, types);
}; // creates the UpdateInput type (not all fields are required)


const updateInputSDL = (model, types = {}) => {
  return inputSDL(model, false, types);
};

const idType = model => {
  var _context4;

  const idField = (0, _find.default)(_context4 = model.fields).call(_context4, field => field.isId);

  if (!idField) {
    throw new Error('Cannot generate SDL without an `id` database column');
  }

  return idField.type;
};

const sdlFromSchemaModel = async name => {
  const model = await (0, _lib.getSchema)(name);

  if (model) {
    var _context5, _context6, _context7, _context8, _context9, _context10;

    // get models for user-defined types referenced
    const types = (0, _reduce.default)(_context5 = await _promise.default.all((0, _map.default)(_context6 = (0, _filter.default)(_context7 = model.fields).call(_context7, field => field.kind === 'object')).call(_context6, async field => {
      const model = await (0, _lib.getSchema)(field.type);
      return model;
    }))).call(_context5, (acc, cur) => ({ ...acc,
      [cur.name]: cur
    }), {}); // Get enum definition and fields from user-defined types

    const enums = (0, _reduce.default)(_context8 = await _promise.default.all((0, _map.default)(_context9 = (0, _filter.default)(_context10 = model.fields).call(_context10, field => field.kind === 'enum')).call(_context9, async field => {
      const enumDef = await (0, _lib.getEnum)(field.type);
      return enumDef;
    }))).call(_context8, (acc, curr) => (0, _concat.default)(acc).call(acc, curr), []);
    return {
      query: querySDL(model).join('\n    '),
      createInput: createInputSDL(model, types).join('\n    '),
      updateInput: updateInputSDL(model, types).join('\n    '),
      idType: idType(model),
      relations: (0, _helpers.relationsForModel)(model),
      enums
    };
  } else {
    throw new Error(`"${name}" model not found, check if it exists in "./api/prisma/schema.prisma"`);
  }
};

const files = async ({
  name,
  crud
}) => {
  const {
    query,
    createInput,
    updateInput,
    idType,
    relations,
    enums
  } = await sdlFromSchemaModel((0, _pascalcase.default)(_pluralize.default.singular(name)));
  const template = (0, _lib.generateTemplate)(_path.default.join('sdl', 'templates', 'sdl.js.template'), {
    name,
    crud,
    query,
    createInput,
    updateInput,
    idType,
    enums
  });

  const outputPath = _path.default.join((0, _lib.getPaths)().api.graphql, `${(0, _camelcase.default)((0, _pluralize.default)(name))}.sdl.js`);

  return {
    [outputPath]: template,
    ...(await (0, _service.files)({
      name,
      crud,
      relations
    }))
  };
};

exports.files = files;
const command = 'sdl <model>';
exports.command = command;
const desc = 'Generate a GraphQL schema and service object.';
exports.desc = desc;
const builder = {
  services: {
    type: 'boolean',
    default: true
  },
  crud: {
    type: 'boolean',
    default: false
  },
  force: {
    type: 'boolean',
    default: false
  }
}; // TODO: Add --dry-run command

exports.builder = builder;

const handler = async ({
  model,
  crud,
  force
}) => {
  var _context11;

  const tasks = new _listr.default((0, _filter.default)(_context11 = [{
    title: 'Generating SDL files...',
    task: async () => {
      const f = await files({
        name: model,
        crud
      });
      return (0, _lib.writeFilesTask)(f, {
        overwriteExisting: force
      });
    }
  }]).call(_context11, Boolean), {
    collapse: false,
    exitOnError: true
  });

  try {
    await tasks.run();
  } catch (e) {
    console.log(_colors.default.error(e.message));
  }
};

exports.handler = handler;
