// @redwoodjs/cli/dist/commands/build.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handler = exports.builder = exports.desc = exports.command = void 0;

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/map"));

var _includes = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/includes"));

var _fs = _interopRequireDefault(require("fs"));

var _path = _interopRequireDefault(require("path"));

var _execa = _interopRequireDefault(require("execa"));

var _listr = _interopRequireDefault(require("listr"));

var _listrVerboseRenderer = _interopRequireDefault(require("listr-verbose-renderer"));

var _lib = require("../lib");

var _colors = _interopRequireDefault(require("../lib/colors"));

var _generate = require("./dbCommands/generate");

const apiExists = _fs.default.existsSync((0, _lib.getPaths)().api.src);

const webExists = _fs.default.existsSync((0, _lib.getPaths)().web.src);

const optionDefault = (webExists, apiExists) => {
  let options = [];
  if (webExists) options.push('web');
  if (apiExists) options.push('api');
  return options;
};

const command = 'build [app..]';
exports.command = command;
const desc = 'Build for production.';
exports.desc = desc;
const builder = {
  app: {
    choices: ['api', 'web'],
    default: optionDefault(webExists, apiExists)
  },
  verbose: {
    type: 'boolean',
    default: false,
    alias: ['v']
  },
  stats: {
    type: 'boolean',
    default: false
  }
};
exports.builder = builder;

const handler = async ({
  app = ['api', 'web'],
  verbose = false,
  stats = false
}) => {
  if ((0, _includes.default)(app).call(app, 'api')) {
    try {
      await (0, _generate.handler)({
        verbose,
        force: true
      });
    } catch (e) {
      console.log(_colors.default.error(e.message));
      process.exit(1);
    }
  }

  const execCommandsForApps = {
    api: {
      // must use path.join() here, and for 'web' below, to support Windows
      cwd: _path.default.join((0, _lib.getPaths)().base, 'api'),
      cmd: 'yarn cross-env NODE_ENV=production babel src --out-dir dist'
    },
    web: {
      cwd: _path.default.join((0, _lib.getPaths)().base, 'web'),
      cmd: `yarn webpack --config ../node_modules/@redwoodjs/core/config/webpack.${stats ? 'stats' : 'production'}.js`
    }
  };

  if (stats) {
    app = ['web'];
    console.log(' â© Skipping `api` build because stats only works for Webpack at the moment.');
  }

  const tasks = new _listr.default((0, _map.default)(app).call(app, appName => {
    const {
      cwd,
      cmd
    } = execCommandsForApps[appName];
    return {
      title: `Building "${appName}"${stats && ' for stats' || ''}...`,
      task: () => {
        return (0, _execa.default)(cmd, undefined, {
          stdio: verbose ? 'inherit' : 'pipe',
          shell: true,
          cwd
        });
      }
    };
  }), {
    renderer: verbose && _listrVerboseRenderer.default
  });

  try {
    await tasks.run();
  } catch (e) {
    console.log(_colors.default.error(e.message));
    process.exit(1);
  }
};

exports.handler = handler;
