// @prisma/engine-core/dist/h1client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.H1Client = void 0;
const http_1 = __importDefault(require("http"));
class H1Client {
    constructor() {
        this.closed = false;
        this.agent = new http_1.default.Agent({ keepAlive: true, maxSockets: 100 });
    }
    request(port, body) {
        return new Promise((resolve, reject) => {
            const req = http_1.default.request({
                agent: this.agent,
                hostname: 'localhost',
                path: '/',
                method: 'POST',
                port,
                headers: {
                    'Content-Type': 'application/json',
                    'Content-Length': Buffer.byteLength(body),
                },
            }, (res) => {
                const chunks = [];
                res.on('data', (chunk) => {
                    chunks.push(chunk);
                });
                res.on('end', () => {
                    resolve({
                        data: JSON.parse(Buffer.concat(chunks)),
                        headers: res.headers,
                    });
                });
            });
            req.on('error', reject);
            req.write(body);
            req.end();
        });
    }
    close() {
        this.agent.destroy();
        this.closed = true;
    }
}
exports.H1Client = H1Client;
