// @redwoodjs/cli/dist/commands/generate/layout/layout.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handler = exports.builder = exports.desc = exports.command = exports.files = void 0;

var _reduce = _interopRequireDefault(require("@babel/runtime-corejs3/core-js/instance/reduce"));

var _helpers = require("../helpers");

const COMPONENT_SUFFIX = 'Layout';
const REDWOOD_WEB_PATH_NAME = 'layouts';

const files = ({
  name
}) => {
  var _context;

  const layoutFile = (0, _helpers.templateForComponentFile)({
    name,
    suffix: COMPONENT_SUFFIX,
    webPathSection: REDWOOD_WEB_PATH_NAME,
    generator: 'layout',
    templatePath: 'layout.js.template'
  });
  const testFile = (0, _helpers.templateForComponentFile)({
    name,
    suffix: COMPONENT_SUFFIX,
    extension: '.test.js',
    webPathSection: REDWOOD_WEB_PATH_NAME,
    generator: 'layout',
    templatePath: 'test.js.template'
  }); // Returns
  // {
  //    "path/to/fileA": "<<<template>>>",
  //    "path/to/fileB": "<<<template>>>",
  // }

  return (0, _reduce.default)(_context = [layoutFile, testFile]).call(_context, (acc, [outputPath, content]) => {
    return {
      [outputPath]: content,
      ...acc
    };
  }, {});
};

exports.files = files;
const {
  command,
  desc,
  builder,
  handler
} = (0, _helpers.createYargsForComponentGeneration)({
  componentName: 'layout',
  filesFn: files
});
exports.handler = handler;
exports.builder = builder;
exports.desc = desc;
exports.command = command;
