// @redwoodjs/api/dist/makeMergedSchema/rootSchema.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.resolvers = exports.schema = void 0;

var _apolloServerLambda = require("apollo-server-lambda");

var _graphqlIsoDate = require("graphql-iso-date");

var _package = _interopRequireDefault(require("../../package.json"));

/**
 * This adds scalar types for dealing with Date, Time, and DateTime,
 * and adds a root Query type which is needed to start the GraphQL server on a
 * fresh install.
 */
const schema = (0, _apolloServerLambda.gql)`
  scalar Date
  scalar Time
  scalar DateTime

  type Redwood {
    version: String
  }

  type Query {
    redwood: Redwood
  }
`;
exports.schema = schema;
const resolvers = {
  Date: _graphqlIsoDate.GraphQLDate,
  Time: _graphqlIsoDate.GraphQLTime,
  DateTime: _graphqlIsoDate.GraphQLDateTime,
  Query: {
    redwood: () => ({
      version: _package.default.version
    })
  }
};
exports.resolvers = resolvers;
